File: frontend/src/lib/api/types/game.ts
import type { Pagination } from "./api";

/**
 * Payload for creating a new game.
 */
export interface CreateGamePayload {
  session_id?: string;
  session_name?: string;
}

/**
 * Response for a created game.
 */
export interface GameResponse {
  game_id: string;
  session_id?: string;
  status: string;
  created_at: string;
  created_by_user_id: string;
  current_scorekeeper_user_id?: string;
}

/**
 * Payload for adding a player to a game.
 */
export interface AddPlayerToGamePayload {
  user_id?: string;
  guest_name?: string;
  seating_order: number;
}

/**
 * Response for an added game player.
 */
export interface GamePlayerResponse {
  game_player_id: string;
  game_id: string;
  user_id?: string;
  guest_player_id?: string;
  display_name: string;
  username?: string;
  avatar_url?: string;
  seating_order: number;
  final_score: number;
}

/**
 * Payload for updating game setings.
 */
export interface UpdateGameSettingsPayload {
  scorekeeper_user_id: string;
  ordered_player_ids: string[];
  starting_dealer_game_player_id: string;
}

/**
 * Holds the score details for a single player in a single round.
 */
export interface PlayerRoundData {
  game_player_id: string;
  bid_amount: number | null;
  tricks_taken: number | null;
  round_score: number | null;
  bonus_points: number | null;
}

/**
 *  Contains all player data for a single round of the scorecard.
 */
export interface RoundScorecard {
  round_number: number;
  status: "bidding" | "playing" | "completed";
  player_scores: PlayerRoundData[];
  dealer_game_player_id: string;
}

/**
 * The main object for fetching and rendering the entire scorecard.
 */
export interface ScorecardResponse {
  game_id: string;
  game_status: string;
  players: GamePlayerResponse[];
  rounds: RoundScorecard[];
  current_round: number;
  session_name?: string | null;
  current_scorekeeper_user_id: string;
  scorekeeper_name: string;
  asterisks: PlayerGameAsterisk[];
}

/**
 * Defines the structure for a single player's bid submission.
 */
export interface PlayerBid {
  game_player_id: string;
  bid_amount: number;
}

/**
 * Payload for submitting all bids for a round.
 */
export interface SubmitBidsPayload {
  bids: PlayerBid[];
}

/**
 * Defines the structure for a single player's trick submission.
 */
export interface PlayerTricks {
  game_player_id: string;
  tricks_taken: number;
  bonus_points: number;
}

/**
 * Payload for submitting all tricks taken for a round.
 */
export interface SubmitTricksPayload {
  tricks: PlayerTricks[];
}

/**
 * Represents a single player with an active game card for the dashboard.
 */
export interface ActiveGamePlayer {
  display_name: string;
  avatar_url?: string | null;
}

/**
 * Represents a single active game card for the dashboard list.
 */
export interface ActiveGameResponse {
  game_id: string;
  session_name?: string | null;
  scorekeeper_name: string;
  is_scorekeeper: boolean;
  created_at: string;
  current_round: number;
  players: ActiveGamePlayer[];
}

export interface GameHistoryItem {
  game_id: string;
  session_name?: string | null;
  game_date: string;
  finishing_position: number;
  total_points: number;
  rounds_hit: number;
  zero_differential: number;
  total_players: number;
  total_asterisks: number;
  scorekeeper_name: string;
}

export interface PaginatedGameHistoryResponse {
  games: GameHistoryItem[];
  pagination: Pagination;
}

/**
 * Payload for adding an asterisk to a player.
 */
export interface AddAsteriskPayload {
  reason: string;
}

/**
 * Represents a single player asterisk.
 */
export interface PlayerGameAsterisk {
  player_game_asterisk_id: string;
  game_player_id: string;
  reason?: string | null;
  created_at: string;
}

File: backend/internal/models/api/game.go
package models

import "time"

// --- Request Payloads ---

// Payload to create a new game
type CreateGameRequest struct {
	SessionID   *string `json:"session_id,omitempty"`
	SessionName *string `json:"session_name,omitempty"`
}

// Request to add a player to a game
type AddPlayerToGameRequest struct {
	UserID       *string `json:"user_id,omitempty"`
	GuestName    *string `json:"guest_name,omitempty"`
	SeatingOrder int     `json:"seating_order" validate:"required,gt=0"`
}

// Request to updating pre-game settings
type UpdateGameSettingsRequest struct {
	ScorekeeperUserID          string   `json:"scorekeeper_user_id" validate:"required"`
	OrderedPlayerIDs           []string `json:"ordered_player_ids" validate:"required"`
	StartingDealerGamePlayerID string   `json:"starting_dealer_game_player_id" validate:"required"`
}

// Request for submitting all bids for a round
type SubmitBidsRequest struct {
	Bids []PlayerBid `json:"bids" validate:"required,min=1,dive"`
}

// Request for submitting all tricks taken for a round
type SubmitTricksRequest struct {
	Tricks []PlayerTrickData `json:"tricks" validate:"required,min=1,dive"`
}

// Request for adding an asterisk to a player
type AddAsteriskRequest struct {
	Reason string `json:"reason"`
}

// --- Response Payloads ---

// Response for a created game
type GameResponse struct {
	GameID                   string    `json:"game_id"`
	SessionID                *string   `json:"session_id,omitempty"`
	Status                   string    `json:"status"`
	CreatedAt                time.Time `json:"created_at"`
	CreatedByUserID          string    `json:"created_by_user_id"`
	CurrentScoreKeeperUserID *string   `json:"current_scorekeeper_user_id,omitempty"`
}

// Represents a single active game for the dashboard list
type ScorecardResponse struct {
	GameID                   string               `json:"game_id"`
	GameStatus               string               `json:"game_status"`
	Players                  []GamePlayerResponse `json:"players"`
	Rounds                   []RoundScorecard     `json:"rounds"`
	CurrentRound             int                  `json:"current_round"`
	SessionName              *string              `json:"session_name,omitempty"`
	CurrentScoreKeeperUserID string               `json:"current_scorekeeper_user_id"`
	ScorekeeperName          string               `json:"scorekeeper_name"`
	Asterisks                []PlayerGameAsterisk `json:"asterisks"`
}

// Represents a single active game for the dashboard list
type ActiveGameResponse struct {
	GameID          string             `json:"game_id"`
	SessionName     *string            `json:"session_name,omitempty"`
	ScorekeeperName string             `json:"scorekeeper_name"`
	IsScorekeeper   bool               `json:"is_scorekeeper"`
	CreatedAt       time.Time          `json:"created_at"`
	CurrentRound    int                `json:"current_round"`
	Players         []ActiveGamePlayer `json:"players"`
}

// Represents the paginated response for the game history
type PaginatedGameHistoryResponse struct {
	Games      []GameHistoryItem `json:"games"`
	Pagination Pagination        `json:"pagination"`
}

// Response for the end-of-game summary screen
type GameSummaryResponse struct {
	WinnerName  string               `json:"winner_name"`
	FinalScores []GamePlayerResponse `json:"final_scores"`
	Awards      []GameAward          `json:"awards"`
}

// --- Component Structs ---

// Represents a single player's bid submission
type PlayerBid struct {
	GamePlayerID string `json:"game_player_id" validate:"required"`
	BidAmount    int    `json:"bid_amount" validate:"gte=0"`
}

// Represents the tricks and bonus points for a player in a round
type PlayerTrickData struct {
	GamePlayerID string `json:"game_player_id" validate:"required"`
	TricksTaken  int    `json:"tricks_taken" validate:"gte=0"`
	BonusPoints  int    `json:"bonus_points" validate:"gte=0"`
}

// Represents a player's details within a game context
type GamePlayerResponse struct {
	GamePlayerID  string  `json:"game_player_id"`
	GameID        string  `json:"game_id"`
	UserID        *string `json:"user_id,omitempty"`
	GuestPlayerID *string `json:"guest_player_id,omitempty"`
	DisplayName   string  `json:"display_name"`
	Username      *string `json:"username,omitempty"`
	AvatarURL     *string `json:"avatar_url,omitempty"`
	SeatingOrder  int     `json:"seating_order"`
	FinalScore    int     `json:"final_score"`
}

// Contains the scoring data for all players for a single round
type RoundScorecard struct {
	RoundNumber        int               `json:"round_number"`
	Status             string            `json:"status"`
	PlayerScores       []PlayerRoundData `json:"player_scores"`
	DealerGamePlayerID string            `json:"dealer_game_player_id"`
}

// Contains an individual player's data for one round on the scorecard
type PlayerRoundData struct {
	GamePlayerID string `json:"game_player_id"`
	BidAmount    *int   `json:"bid_amount,omitempty"`
	TricksTaken  *int   `json:"tricks_taken,omitempty"`
	RoundScore   *int   `json:"round_score,omitempty"`
	BonusPoints  *int   `json:"bonus_points,omitempty"`
}

// Represents a single player within an active game card on the dashboard
type ActiveGamePlayer struct {
	DisplayName string  `json:"display_name"`
	AvatarURL   *string `json:"avatar_url,omitempty"`
}

// Represents an asterisk given to a player in a game
type PlayerGameAsterisk struct {
	PlayerGameAsteriskID string    `json:"player_game_asterisk_id"`
	GamePlayerID         string    `json:"game_player_id"`
	Reason               *string   `json:"reason,omitempty"`
	CreatedAt            time.Time `json:"created_at"`
}

// Represents a game award
type GameHistoryAward struct {
	Type  string `json:"type"`
	Title string `json:"title"`
}

// Represents a single row in the user's game history table
type GameHistoryItem struct {
	GameID            string             `json:"game_id"`
	SessionName       *string            `json:"session_name,omitempty"`
	GameDate          time.Time          `json:"game_date"`
	FinishingPosition int                `json:"finishing_position"`
	TotalPoints       int                `json:"total_points"`
	RoundsHit         int                `json:"rounds_hit"`
	ZeroDifferential  int                `json:"zero_differential"`
	TotalPlayers      int                `json:"total_players"`
	TotalAsterisks    int                `json:"total_asterisks"`
	ScorekeeperName   string             `json:"scorekeeper_name"`
	AwardsWon         []GameHistoryAward `json:"awards_won,omitempty"`
}

// Represents a special award given at the end of a game
type GameAward struct {
	Title       string `json:"title"`
	PlayerName  string `json:"player_name"`
	Value       string `json:"value"`
	Description string `json:"description"`
}
